pkg/util/writerlease/writerlease.go

```diff
@@ -0,0 +1,295 @@
+func NewWithBackoff(name string, leaseDuration, retryInterval time.Duration, backoff wait.Backoff) *WriterLease {
+	return &WriterLease{
+		name:          name,
+		backoff:       backoff,
+		maxBackoff:    leaseDuration,
+		retryInterval: retryInterval,
+
+		queued: make(map[string]WorkFunc),
+		queue:  workqueue.NewNamedDelayingQueue(name),
+		once:   make(chan struct{}),
+	}
+}
+
+func (l *WriterLease) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer l.queue.ShutDown()
+
+	go func() {
+		defer utilruntime.HandleCrash()
+		l.work()
+		glog.V(4).Infof("[%s] Worker stopped", l.name)
+	}()
+
+	<-stopCh
+}
+
+func (l *WriterLease) Expire() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	l.expires = time.Time{}
+}
+
+func (l *WriterLease) Wait() bool {
+	<-l.once
+	state, _, _ := l.leaseState()
+	return state == Leader
+}
+
+func (l *WriterLease) WaitUntil(t time.Duration) (bool, bool) {
+	select {
+	case <-l.once:
+	case <-time.After(t):
+		return false, false
+	}
+	state, _, _ := l.leaseState()
+	return state == Leader, true
+}
+
+func (l *WriterLease) Try(key string, fn WorkFunc) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	l.queued[key] = fn
+	if l.state == Follower {
+		l.queue.AddAfter(key, l.expires.Sub(nowFn()))
+	} else {
+		l.queue.Add(key)
+	}
+}
+
+func (l *WriterLease) Extend(key string) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	if _, ok := l.queued[key]; ok {
+		delete(l.queued, key)
+		switch l.state {
+		case Follower:
+			l.tick++
+			backoff := l.nextBackoff()
+			glog.V(4).Infof("[%s] Clearing work for %s and extending lease by %s", l.name, key, backoff)
+			l.expires = nowFn().Add(backoff)
+		}
+	}
+}
+
+func (l *WriterLease) Len() int {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	return len(l.queued)
+}
+
+func (l *WriterLease) Remove(key string) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	delete(l.queued, key)
+}
+
+func (l *WriterLease) get(key string) WorkFunc {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	return l.queued[key]
+}
+
+func (l *WriterLease) leaseState() (State, time.Time, int) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	return l.state, l.expires, l.tick
+}
+
+func (l *WriterLease) work() {
+	for {
+		item, shutdown := l.queue.Get()
+		if shutdown {
+			return
+		}
+		key := item.(string)
+
+		fn := l.get(key)
+		if fn == nil {
+			glog.V(4).Infof("[%s] Work item %s was cleared, done", l.name, key)
+			l.queue.Done(key)
+			continue
+		}
+
+		leaseState, leaseExpires, _ := l.leaseState()
+		if leaseState == Follower {
+			// if we are following, continue to defer work until the lease expires
+			if remaining := leaseExpires.Sub(nowFn()); remaining > 0 {
+				glog.V(4).Infof("[%s] Follower, %s remaining in lease", l.name, remaining)
+				l.queue.AddAfter(key, remaining)
+				l.queue.Done(key)
+				continue
+			}
+			glog.V(4).Infof("[%s] Lease expired, running %s", l.name, key)
+		} else {
+			glog.V(4).Infof("[%s] Lease owner or electing, running %s", l.name, key)
+		}
+
+		isLeader, retry := fn()
```

Comment:
Not sure I `grok` this part. So we call the work queue function (in this case to  update status) -  https://github.com/smarterclayton/origin/blob/b1f1f97f0eee4ae5325a05731288ece94e864f38/pkg/router/controller/status.go#L252  and its response says its the leader? Aren't we already the leader here if we got to this part.

Reply:
This is a lease that is driven by work renewal.  If a client observes that no work has been done within the lease window, it can compete to acquire the lease by doing work.

If we succeed (return true for the work), then we have "acquired the lease" by virtue of doing the work.  The route object status itself is acting as the lease.